src/main/java/com/cloudbees/plugins/credentials/CredentialsProvider.java

```diff
@@ -887,32 +887,33 @@ public void remove() {
                     isParameter = true;
                     isDefaultValue = ((CredentialsParameterValue) parameter).isDefaultValue();
                     id = ((CredentialsParameterValue) parameter).getValue();
+                    if (id == null)
+                        return null;
                 }
             }
         }
+        List<C> jobContextCandidates = new ArrayList<>();
         // non parameters or default parameter values can only come from the job's context
         if (!isParameter || isDefaultValue) {
             // we use the default authentication of the job as those are the only ones that can be configured
             // if a different strategy is in play it doesn't make sense to consider the run-time authentication
             // as you would have no way to configure it
             Authentication runAuth = CredentialsProvider.getDefaultAuthenticationOf(run.getParent());
-            List<C> candidates = new ArrayList<C>();
             // we want the credentials available to the user the build is running as
-            candidates.addAll(
+            jobContextCandidates.addAll(
                     CredentialsProvider.lookupCredentials(type, run.getParent(), runAuth, domainRequirements)
             );
             // if that user can use the item's credentials, add those in too
             if (runAuth != ACL.SYSTEM && run.getACL().hasPermission(runAuth, CredentialsProvider.USE_ITEM)) {
-                candidates.addAll(
+                jobContextCandidates.addAll(
                         CredentialsProvider.lookupCredentials(type, run.getParent(), ACL.SYSTEM, domainRequirements)
                 );
             }
-            return CredentialsMatchers.firstOrNull(candidates, CredentialsMatchers.withId(id));
```

Comment:
I'm getting scared by this.

The intent (as documented by the comment on line 893) is that non-parameters and default parameter values can only come from the job's context and not from the user's context.

Now it may be that we need to revisit that... but there are implications in opening it up.